### PR TITLE
Try to figure out where/how to insert the cache here

### LINKED ISSUE #730: Enable OwnedPaths (i.e. FileFormat and DirectoryFormat with write mode) to initialize in a Cache
**Addition Description**
This feature is largely for accelerating importing by avoiding unnecessary disk IO. 

**Current Behavior**
Currently creating an OwnedPath in QIIME 2 will use $TMPDIR which can be problematic if the destination is ultimately a Cache on a different block device. This will result in a copy of all of the data.

**Proposed Behavior**
We should have a mechanism to cause any instantiated OwnedPaths to appear in a target location. Plugin code/transformers should not be aware of this, so the immediate call-site will not control this. However the originating call-site (such as the Artifact/Python 3 API) should have a way to control it.

### PR DESCRIPTION
**Pull request by Oddant1**
Closes #730

### CODE REVIEW

**Comment by Oddant1** [comment]
@ebolyen is this along the lines of what we want? It seems to be doing what we wanted. This isn't the part of the framework I'm most familiar with.

**Comment by ebolyen** [comment]
Almost, I would lift this logic up to `qiime2.core.path`.

Is the process pool the best place? It seems kind of reasonable, although it's looking less like a pool every day. The automated cleanup is a nice feature of it.

We should also test this with your parsl situation with the provenance path not showing up. My hunch is it will still work since there's a flufl lock, but worth confirming before we break everything.

**Comment by Oddant1** [comment]
Yeah while we were messing with that I was wondering if this functionality would resolve that

**Comment by Oddant1** [comment]
Ok so we do want to put this in `__new__` for OwnedPath or something?

**Comment by Oddant1** [comment]
Something like how it is now and stop passing in prefix? I had figured this would bork things pretty bad, but maybe not.

**Comment by ebolyen** [comment]
Yep, in theory that's all we should need to do.

File: qiime2/core/path.py [hunk 1]
```diff
@@ -24,6 +24,12 @@ def _party_parrot(self, *args):
 
 class OwnedPath(_ConcretePath):
     def __new__(cls, *args, **kwargs):
+        from qiime2.core.cache import get_cache
+
+        cache = get_cache()
+        kwargs['prefix'] = os.path.join(cache.process_pool.path,
+                                        'q2-%s-' % cls.__name__)
```
Comment (ebolyen) [inline qiime2/core/path.py]: Let's put this under a `tmp/` directory in the process pool (creating if needed). Maybe there should be a helper on the cache itself to return the temp directory?

**Comment by Oddant1** [comment]
This does not work how I thought it did at all. The actual path is one of the args, and the prefix doesn't seem to do much if anything when passed up to pathlib.Path

**Comment by Oddant1** [comment]
@ebolyen I think putting the cache usage in `OutPath.__new__` does something closer to what you wanted.

File: qiime2/core/path.py [hunk 2]
```diff
@@ -92,14 +92,20 @@ def _destruct(cls, path):
         else:
             os.unlink(path)
 
-    def __new__(cls, dir=False, **kwargs):
```
Comment (ebolyen) [inline qiime2/core/path.py:95]: Where there any callers of this? I swear we set `prefix` in a lot of places for stuff like `q2-provenance`... although come to think of it, cache has probably eroded out most of those callsites, but there may still be a few left.

Comment (Oddant1) [reply qiime2/core/path.py:95]: Only existing use is https://github.com/qiime2/qiime2/blob/c95b12bde54170b18a80c2c95f13decfb893ee14/qiime2/core/format.py#L23 and tests

**Comment by Oddant1** [comment]
The child processes will not clean these tmp dirs out when they exit because their termination does not call exit funcs.

**Comment by Oddant1** [comment]
This may actually be useful for #790

**Comment by ebolyen** [comment]
Superseded by #790

**Pull request closed without merging**